### ISSUE TITLE
The beer problem. Need bettter name for resource.

### ISSUE THREAD

**Issue by Moschn** [amiv-eth/amivapi#129]
After each semester the AMIV treasurer has to bill the other student associations for the amount of beers consumed by them. As far as I know these purchases cannot yet be logged correctly. Is there a membership value for VIS or VMP?

A solution to log all beers could look like this:
The "beer" client posts not only the purchase from our members but also the purchases from the other student organisations. These purchases are linked to users with some specific membership value (e.g. "VIS").  If all purchases are logged then it is trivial to find the beer amount for every student association. 

The new user creation can be done on the beer client, e.g. if the user does not yet exist, request the legi_nr from VIS/VMP and then create a new user.

**Comment by fubu** [comment]
I don't think that should be handled by our API, but rather by the beer machine/service itself - it has all the necessary data available, for free, already.

Note: The initial design of the beer service on purpose only let the relevant student association know about purchases, as the others - as you described - don't (have to) know about other users.

**Comment by Moschn** [comment]
At the beginning I also thought it can be handled by the beer machine, but there are a few disadvantages:
- It is harder to access. Lets think about the treasurer in 5 years. He has to be able to log into the beer machine and reset (?) the counter and extract the data. It would be easier if this was easily available from the AMIV website for example.
- The API is fail safe. The beer machine not at all. It will be hard to bill if we have no data. I do not think we should store any relevant data on the beer machine.  

How can students from other studen associations access the study documents? This is a similar question, and if this is possible the same thing is easily imp,emented for beer purchases. 

What are the advantages of storing this data on the beer machine itself?

**Comment by NotSpecial** [comment]
I don't really know a lot about that issue, just a few thoughts:

* Access: The beer machine website can use the amivapi to let the quästor log in. e.g. [this little website](https://afternoondelight.de/contractor) to help the contact team. Login also done via api.
To add more security one could also require the person logging in to be in a group, e.g. "beeradmin", to access the data (and just write this on the login screen)

* Other advantages: The smaller the code of the api is, the better, I think. So if something is not really needed I would be pro removing it.

* Advantage of including it: We already have the "purchase" resource (although I am a little critical about that, see the point above, but well) and it would be rather easy to include it I think

**Comment by fubu** [comment]
Why would the treasurer need to "log into the beer machine and reset" anything? What the treasurer needs, and what the beer service should provide, is a simple admin interface to control/inspect the state of the machine, which could also host the display of usage data. The treasurer could go there and filter the data by his needs. (You are right that the beer service currently does not have that - I think that should be the next step there.)

Also, note that I made a distinction between the beer *service* and the beer *machine* - nobody said that the former only has to include (and run solely on) the latter. I think we can be as fail-safe as the API by e.g. utilizing the MySQL/hosting service provided by the ISG.EE (with automatic backups etc).

My main point is that the beer service, IMHO, is not one of the core AMIV services we want to support in the API itself. What we do support is the "beer rules" component (which determines the amount of free beers left by looking at the vend history), but that's because we want the API to support the "beer/coffee machine protocol" of these other services (as *our members*, which the API serves first ond foremost, demand this).

About students of other associations accessing our study documents: I don't see why this is the same thing. StudyDocuments is a first-class resource of the API (and owned by it) - AMIV members produce them, AMIV members use them, AMIV members decide to share them with others. But other people's beer consumption is not a core functionality of the API. One of the goals of the API was to keep its scope as narrow as possible, while encouraging others to build their own "microservices" on top of it (like a PVK tool). For me, the beer service is a clear and natural boundary where we can seperate responsabilities. The API team should not have to support the beer machine (it's just not that important in the grand scheme of AMIV things), but rather keep it autonomous and open for other people to work on it without jeopardizing the "core" API.

**Comment by NotSpecial** [comment]
@fubu, I like your thoughts on the *purchase* resource, I haven't thought of it in that light. Seeing it as an implementation of a *beer protocol* makes much more sense. I take back my critique. Maybe we should really rename it to make that clearer. Suggestions? ;)

**Comment by Moschn** [comment]
I understand your points. I would suggest to remove the purchase resource entirely. It is misleading to have such a resource if it only supports a part of the "beer protocol". We will then write another application which can handle the coffee and beer vending machines. This would help to keep the API simple. 

@fubu: I only talked about the StudyDocuments because I do not know how other user are handled (I guess no their membership value is set to "none"?).

**Comment by NotSpecial** [comment]
I am not sure if this is what @fubu wanted to say.

I think he wants to make a separation of the *beer protocol*, which represents the rule:

> Every AMIV member can get a beer or coffee every day

From the actual service providing it. There could be even 100 concurrent services taking care of the beer and coffee thing. But as long as they all ask the API:

> Can user *Pablo* have a beer?

Everything will be going on according to our rule. This has some benefits, I think:

- The beer service doesn't have to be maintained by the api team
- Beer and coffee service can be separate if needed
- New (or substitute) services can be introduced easily

Since the beer service might change more than the API, I think it makes sense to keep the core in the API, though. This way the VIS coffee machine can directly communicate with our API only.

Otherwise a change in the beer service would also kill the coffee access, which is not nice.

Furthermore I think this makes implementing a beer service quite easy, since it can treat AMIV beer access the same way as VIS or VMP beer access:
```
if AMIV member: query AMIVAPI
if VIS member: query VIS-equivalent
if ...
```

But I'm talking too much. How is communication with VIS/VMP handled by the beer machine currently?


Furthermore I think I might be missing something. Regarding:

>if it only supports a part of the "beer protocol"

What part of the beer protocol would not be supported, @Moschn ?

**Comment by Moschn** [comment]
Ok I will try to explain it again:

The beer machine queries all possible APIs from different student associations. The response contains the amount of beers available. If the vending of the beer is complete, this purchase has to be written in some log. We need to somehow log the purchases on order to bill the VIS/VMP at the end of the semester.  Not only the purchase from AMIV users but also the ones from VIS and VMP users have to be written to this log. If we handle this log seperate from the API we have to still notify the API that a beer has been purchased and then notify the beer "microservice" the same thing.  For me it makes no sense to store this information in two services at the same time. A such "microservice" can easily handle the purchase resource by itself. It just needs the amount of beers per user per day.

We also have to have a different user management in the "microservice" because the API does not store any value indicating this user is from VIS/VMP or AMIV but this is a problem for another time.

**Comment by fubu** [comment]
TL;DR: I still think it is better to not to intermix the general beer service data with the AMIV API. I present my proposal for the beer service in the end of this comment. Everything in between is some explanation of how the "beer protocol"/service was designed for.


So let's clear up some confusion.

# The beer protocol

> I think he wants to make a separation of the beer protocol, which represents the rule:
> 
> > Every AMIV member can get a beer or coffee every day

That's not the "beer protocol", that's just the implementation AMIV chose. The beer protocol only says that:

1. Any participating organization ("identity provider") needs to provide a way to authenticate the RFID card numbers it wants to represent.
2. The identity provider also needs to allow or prohibit specific requests to give out a free beer. (For efficiency reasons this is usually combined with the response to (1).)
3. The identity provider who grants such a request for a free beer should also provide a way to be notified about the successful vend request.

After some discussion we decided to support this protocol in the API, as it is the single source of truth for AMIV membership data. For the implementation we went with the previous (and obvious) way of doing it: (1) Have some way to query the `User` resource by RFID card number, (2) have the query response contain some field which represents how many beers a user is eligible to get for free at the time of the request, and (3) have a log of previous successful vend requests (the `Purchase` resource) where the beer service can write to.

Just for completeness, that's more or less also how the coffee machine / "coffee protocol" works.

> From the actual service providing it. There could be even 100 concurrent services taking care of the beer and coffee thing. But as long as they all ask the API:
>
> > Can user Pablo have a beer?
>
> Everything will be going on according to our rule. This has some benefits, I think:
>
> * The beer service doesn't have to be maintained by the api team
> * Beer and coffee service can be separate if needed
> * New (or substitute) services can be introduced easily

This is exactly the reasoning behind the system design outlined above. The beer service should never depend on another system "working fine" (think: AMIV API is down, so no VIS'ler can get beer/coffee or vice versa), but it also should never have to worry about the rules in place (or in other words, the identity providers should always be in charge of deciding on their own rules without having to change the beer service). If nobody is there who wants to give out free beer to the presented legi card, that's not the problem of the beer service. Furthermore, with the original proposal from this issue in mind, if the beer maintainers decide to change something they should not have to go and change other people's code (like the API team's API code). The beer service is supposed to work/be managed independently from the "identity providers".

> Furthermore I think this makes implementing a beer service quite easy, since it can treat AMIV beer access the same way as VIS or VMP beer access:
>
>     if AMIV member: query AMIVAPI
>     if VIS member: query VIS-equivalent
>     if ...

Just a minor correction: The beer service only knows about who you are *after* it queried the identity providers. So it's more like a loop over the registered identity providers:

    ID_PROVIDERS = [AMIV_ID, VIS_ID, ...]
    for id_provider in ID_PROVIDERS:
        id_provider.auth(rfid)

# The problem

So, now to the points raised by @Moschn:

> If the vending of the beer is complete, this purchase has to be written in some log.

I agree.
(Except that I would not call it a *purchase*, as there was never any money involved from the point of view of the user. If anything it's a *sale*, as at least AMIV pays something for it, although it chose to sell it at a loss.)

> We need to somehow log the purchases on order to bill the VIS/VMP at the end of the semester.

I agree.
And FWIW, that log already exists (every transaction, successful or not, is stored in Redis on the beer machine). Not as accessible as you want it to be, but that can be changed (see my proposal below).

> Not only the purchase from AMIV users but also the ones from VIS and VMP users have to be written to this log and this is not supported by the API (@NotSpecial).

This is correct. And I still think it *shouldn't*. See below.

> If we handle this log seperate from the API we have to still notify the API that a beer has been purchased and then notify the beer "microservice" the same thing.

The beer service only notifies the responsible identity provider of a successful vend. That might be AMIV in most cases, but is also VIS or VMP in some. This is not to say that the beer service should not store *its own log*. See my proposal below.

> We also have to have a different user management in the "microservice" because the API does not store any value indicating this user is from VIS/VMP or AMIV but this is a problem for another time.

That's exactly one of the reasons why I'm so against including the complete selling log in the API: We never intended to handle VMP's or VIS's user accounts in the AMIV API. We do support login with any N.ETHZ, but it's not on the API to attribute users to other associations. The entity which most accurately knows about these things is the beer machine/service itself. So let's try to keep it seperate from the API.


# The proposal

I think we have already established that

1. We need a log of *all* transactions handled by the beer machine.
2. The log needs to be accessible to at least the AMIV executives for inspection.
3. The log should not reside on the beer machine itself, for durability reasons.

When I talked about the distinction of a beer *service* as opposed to the beer *machine* in a previous comment, I did not mean to physically seperate these two. We already have a beer service, no need to invent a new one: It is running on (and controlling) the beer machine. It even has a complete log, so we already got (1) covered.

Unfortunately this setup does neither satisfy (2) nor (3): To access it you need to log into the beer machine, and if the beer machine goes down the log goes down with it. To migitate this I propose to extract or duplicate this log on to the ISG.EE-managed servers. There is already a `beerlog` MySQL database, even with two different database users (one to write, one for read-only access). An associated ISG.EE account could then use the read-only access to publish its own data analysis frontend. This frontend could also be used to control the beer machine: Shutting down or restarting the machine from the browser, or disabling free beers, or get some statistics from the beer machine itself, ...

In short:

1. Whenever the beer service registers a successful transaction on the beer machine, it writes the data `(user, id_provider, slot, timestamp)` back to the MySQL database at ISG.EE. (This can go directly through MySQL over the network, or with an intermediate "API" layer via HTTPS - the former works perfectly fine for me.)
2. Whenever the treasurer wants to bill the other associations, he/she logs into the frontend (this could use the API login, for example) and selects the date range he wants to evaluate.

Does that sound reasonable to you?

**Comment by Moschn** [comment]
Yes we can do that. I understand why you want to keep the purchase resource and we will try to implement something like your proposal.

**Comment by NotSpecial** [comment]
Reopened, we still need to improve the resource. Any names that are better than "purchase"?

**Comment by Tar-Tarus** [comment]
I propose the name __beverages__ . What are your opinions?

**Comment by Leonidaz0r** [comment]
fine for me
